server/src/test/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java

```diff
@@ -787,9 +787,11 @@ public void sendRequest(Transport.Connection connection, long requestId, String
                 if (PeerRecoverySourceService.Actions.START_RECOVERY.equals(action) && count.incrementAndGet() == 1) {
                     // ensures that it's considered as valid recovery attempt by source
                     try {
-                        awaitBusy(() -> client(blueNodeName).admin().cluster().prepareState().setLocal(true).get()
-                            .getState().getRoutingTable().index("test").shard(0).getAllInitializingShards().isEmpty() == false);
-                    } catch (InterruptedException e) {
+                        assertBusy(() -> assertFalse(
+                            "Expected there to be some initializing shards",
+                            client(blueNodeName).admin().cluster().prepareState().setLocal(true).get()
+                                .getState().getRoutingTable().index("test").shard(0).getAllInitializingShards().isEmpty()));
+                    } catch (Exception e) {
```

Comment:
So there's a couple of things here:

   1. `assertBusy` accepts a `CheckedRunnable<Exception>`, so from the compiler's point-of-view the method either needs to handle the thrown `Exception` or allow it to propagate.
   2. The method also calls `Thread.sleep()`, which throws an `InterruptedException`.

We could get around (1) by using a `CheckedRunnable<AssertionError>` (and changing how `CheckedRunnable` is defined), though that might be a problem for code that not only performs assertions but also operations that can throw other kinds of exceptions.

That still leaves (2). We could rethrow `InteruptedException` as some kind of runtime exception, but that feels horrible.

Reply:
`assertBusy` will only catch (and thus retry) assertions. The inner loop looks like this:

```
try {
    codeBlock.run();
    return;
} catch (AssertionError e) {
    failures.add(e);
}
```

The reason for the `CheckedRunnable<Exception>` is to allow other errors to escape the retries. That escaping is exactly what I want to preserve here. We shouldn't keep retrying if the test hits eg an illegal state, we should let the entire test fail right then, so no need to catch Exception.

Reply:
This particular piece of code isn't test code _per se_, rather it's a stub interface implementation, declared to only throw `IOException`. Since `assertBusy` declares it throws `Exception`, I have to handle somehow.

Reply:
Since assertBusy is called from test methods, I don't think we need to handle Exception. The point is if a general Exception is thrown, the test should fail, so we just need to have the test method throw Exception.

Reply:
When I said this isn't test code, I meant that this code is part of an implementation of the interface `StubbableTransport.SendRequestBehavior`, specifically:

```java
void sendRequest(Connection connection, long requestId, String action, TransportRequest request, TransportRequestOptions options) throws IOException;
```

So I can't take out the exception because the compiler will complain that I'm not handling `Exception`, and I can't change the method signature without also changing the interface (which I could do, if you think it's worthwhile).